### PR TITLE
PR: Add documentation for the Spyder-Unittest plugin

### PR DESCRIPTION
**Pull request by pkpatricia**
# Pull Request

## Pull Request Checklist

* [x] Read and followed this repo's [Contributing Guidelines](https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md)
* [x] Based your PR on the latest version of the correct branch (master or 4.x)
* [x] Checked your writing carefully for correct English spelling, grammar, etc
* [x] Described your changes and the motivation for them below
* [x] Noted what issue(s) this pull request resolves, creating one if needed

## Issue(s) Resolved

Fixes #

### CODE REVIEW

File: doc/plugins/unittest.rst [hunk 1]
```diff
@@ -0,0 +1,69 @@
+###############
+Spyder Unittest
+###############
+
+**Spyder-unittest** is a plugin that integrates popular unit test frameworks with Spyder, allowing you to run test suites and view the results in the IDE.
+
+.. image:: /images/unittest/unittest-run.png
+   :alt: Spyder Unittest in Spyder
+
+Using unittest inside Spyder allows you to take advantage of their web interface alongside Spyder’s powerful features such as the Variable explorer, console and debugger.
```
Comment (juanis2112) [inline doc/plugins/unittest.rst:10]:
```suggestion
Using unittest inside Spyder allows you to take advantage of their web interface alongside Spyder’s powerful features such as the Variable explorer, console and debugger.
```
This is left from the notebook one and I'm not sure it applies to this one but @ccordoba12 can confirm.

File: doc/plugins/unittest.rst [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+==============
+
+When Unittest is installed, it will be available under the menu item :menuselection:`View --> Panes --> Unit testing`.
+
+.. image:: /images/unittest/unittest-view-panes.png
+   :alt: Spyder showing view panes Unittest
+
+You will see it then as a tab in the bottom of the top right area in the same location as the Variable explorer, Help, and Plots views.
+
+Press on the  :guilabel:`Run Green Arrow` to run the testing.
```
Comment (juanis2112) [inline doc/plugins/unittest.rst:41]:
```suggestion
Press on the  :guilabel:`Run Green Arrow` to run the testing.
```
I think here it might be worth to explain that we need to have some tests written in the editor and maybe mention what a test is in Python with a link to this page where there is more information about it https://docs.python.org/3/library/unittest.html. What do you think @ccordoba12 ?

File: doc/plugins/unittest.rst [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+.. image:: /images/unittest/unittest-run-button.png
+   :alt: Spyder showing Unittest run button
+
+You can get an expanded view of the testing messages by clicking on the expand option.
+
+.. image:: /images/unittest/unittest-expanded-message-view.png
+   :alt: Spyder showing Unittest expanded message view
+
+Press on the  :guilabel:`Hamburger Menu Icon` to view different unittest options.
```
Comment (juanis2112) [inline doc/plugins/unittest.rst:51]: Before this one I think it could be worth adding the "Advanced options" since these options are not that common.

File: doc/plugins/unittest.rst [hunk 4]
```diff
@@ -0,0 +1,69 @@
+###############
+Spyder Unittest
+###############
+
+**Spyder-unittest** is a plugin that integrates popular unit test frameworks with Spyder, allowing you to run test suites and view the results in the IDE.
```
Comment (juanis2112) [inline doc/plugins/unittest.rst:5]: Add link in unit test to useful docs. Maybe this one: https://docs.python.org/3/library/unittest.html but let me know @ccordoba12 if you have something better.

File: doc/plugins/unittest.rst [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+.. image:: /images/unittest/unittest-configuration-view.png
+   :alt: Spyder Unittest configuration view in Spyder
+
+The dependencies view will show you all of your installed package versions.
+
+.. image:: /images/unittest/unittest-dependencies-view.png
+   :alt: Spyder Unittest dependencies view in Spyder
+
+The output viewer can be used to directly view the output results of your testing.  This view can be very useful for viewing the results of unittests.
```
Comment (juanis2112) [inline doc/plugins/unittest.rst:66]:
```suggestion
Finally, the output viewer can be used to directly view the output results of your testing.  This view can be very useful for viewing the results of unittests.
```
To improve the flow

**Review by juanis2112 — COMMENTED** [review]
This looks great. Some minor suggestions but amazing work!

**Comment by juanis2112** [comment]
@jitseniesen we have now docs for our Unittest plugin thanks to @pkpatricia  who worked on it during our Scipy sprints! Please let us know what you think about this PR and if there is something we are missing.

File: doc/plugins/unittest.rst (hunk 4 above)
Comment (ccordoba12) [reply doc/plugins/unittest.rst:5]:
```suggestion
**Spyder-unittest** is a plugin that integrates popular unit test frameworks with Spyder, such as Pytest, Unittest and Nose, allowing you to run test suites and view the results in the IDE.
```

- I agree with @juanis2112 that it'd be a nice idea to add links for the testing frameworks we support.
- That's why I added an explicit mention to them above.

File: doc/plugins/unittest.rst (hunk 1 above)
Comment (ccordoba12) [reply doc/plugins/unittest.rst:10]: Yeah, it doesn't apply here.

File: doc/plugins/unittest.rst [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+Installing Unittest
+===================
+
+.. code-block:: bash
+
+   conda install -c spyder-ide spyder-unittest
+
+.. important::
+
+   Please make sure you have Python's pytest package installed if you are going to use that program as well.
```
Comment (ccordoba12) [inline doc/plugins/unittest.rst:22]:
```suggestion
   Please make sure you have Python's Pytest package installed if you are going to use that program as well.
```

Small improvement.

File: doc/plugins/unittest.rst [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+   conda install pytest
+
+Restart Spyder in order to be able to use the plugin.
+
+==============
+Using Unittest
+==============
+
+When Unittest is installed, it will be available under the menu item :menuselection:`View --> Panes --> Unit testing`.
```
Comment (ccordoba12) [inline doc/plugins/unittest.rst:34]:
```suggestion
When Spyder-unittest is installed, it will be available under the menu item :menuselection:`View --> Panes --> Unit testing`.
```

Unittest is the name of standard library module for testing, so I think it's better to be more explicit here.

File: doc/plugins/unittest.rst [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+.. image:: /images/unittest/unittest-view-panes.png
+   :alt: Spyder showing view panes Unittest
+
+You will see it then as a tab in the bottom of the top right area in the same location as the Variable explorer, Help, and Plots views.
```
Comment (ccordoba12) [inline doc/plugins/unittest.rst:39]:
```suggestion
You will see it then as a tab in the bottom of the top right area in the same location as the Variable explorer, Help, and Plots panes.
```

Small correction.

File: doc/plugins/unittest.rst (hunk 2 above)
Comment (ccordoba12) [reply doc/plugins/unittest.rst:41]: Yeah, I think that's important to explain first, i.e. before clicking `Run`, you need to have at least a file whose name starts by `test_` and then functions on it that also start by `test_`. And below that explanation, we could also show a screenshot showing that in the editor, as @juanis2112 suggested.

Of course, that's if we're going to use Pytest to show the plugin functionality, which I think is much simpler to explain than Unittest.

File: doc/plugins/unittest.rst (hunk 3 above)
Comment (ccordoba12) [reply doc/plugins/unittest.rst:51]: Agreed.

File: doc/plugins/unittest.rst [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+.. image:: /images/unittest/unittest-hamburger-menu.png
+   :alt: Spyder Unittest hamburger menu in Spyder
+
+The configuration view will show you how to choose the pytest option and the directory of where you want to run the test file(s) from.
+
+.. image:: /images/unittest/unittest-configuration-view.png
+   :alt: Spyder Unittest configuration view in Spyder
+
+The dependencies view will show you all of your installed package versions.
```
Comment (ccordoba12) [inline doc/plugins/unittest.rst:61]:
```suggestion
The dependencies view will show you all of your installed testing frameworks.
```

Small correction.

**Review by ccordoba12 — COMMENTED** [review]
Thanks @pkpatricia for your help documenting our testing plugin!

File: doc/plugins/unittest.rst (hunk 1 above)
Comment (pkpatricia) [reply doc/plugins/unittest.rst:10]: this was removed

File: doc/plugins/unittest.rst (hunk 2 above)
Comment (pkpatricia) [reply doc/plugins/unittest.rst:41]: The link to the Pytest website and the inclusion of the requirements for a testing file to run were included.

File: doc/plugins/unittest.rst (hunk 3 above)
Comment (pkpatricia) [reply doc/plugins/unittest.rst:59]: Added information on the configuration view, dependencies view, and the output viewer.

File: doc/plugins/unittest.rst (hunk 4 above)
Comment (pkpatricia) [reply doc/plugins/unittest.rst]: Links to Pytest, Unittest and Nose was added.

File: doc/plugins/unittest.rst (hunk 8 above)
Comment (pkpatricia) [reply doc/plugins/unittest.rst:43]: changed

File: doc/plugins/unittest.rst (hunk 9 above)
Comment (pkpatricia) [reply doc/plugins/unittest.rst:69]: changed.

File: doc/plugins/unittest.rst (hunk 5 above)
Comment (pkpatricia) [reply doc/plugins/unittest.rst]: changed to suggested text.

File: doc/plugins/unittest.rst (hunk 6 above)
Comment (pkpatricia) [reply doc/plugins/unittest.rst]: changed case of letter to upper.

File: doc/plugins/unittest.rst (hunk 7 above)
Comment (pkpatricia) [reply doc/plugins/unittest.rst]: changed.

File: doc/plugins/unittest.rst [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+For more on the Python Pytest package:
+
+| `Pytest Run Guide <https://docs.pytest.org/en/stable/getting-started.html#run-multiple-tests>`_
+
+| `Real Python Article on using Pytest <https://realpython.com/pytest-python-testing/>`_
+
+| `Software Testing Help's Pytest Tutorial <https://www.softwaretestinghelp.com/pytest-tutorial/>`_
+
+.. image:: /images/unittest/unittest-run-button.png
```
Comment (juanis2112) [inline doc/plugins/unittest.rst:57]:
```suggestion
```

File: doc/plugins/unittest.rst [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+   :alt: Spyder showing Unittest run button
```
Comment (juanis2112) [inline doc/plugins/unittest.rst:58]:
```suggestion
```

File: doc/plugins/unittest.rst [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+==============
+
+When Spyder-unittest is installed, it will be available under the menu item :menuselection:`View --> Panes --> Unit testing`.
+
+.. image:: /images/unittest/unittest-view-panes.png
+   :alt: Spyder showing view panes Unittest
+
+You will see it then as a tab in the bottom of the top right area in the same location as the Variable explorer, Help, and Plots panes.
+
+Press on the  :guilabel:`Run Green Arrow` to run the testing.
```
Comment (juanis2112) [inline doc/plugins/unittest.rst:45]:
```suggestion
Press on the  :guilabel:`Run Green Arrow` to run the testing.

.. image:: /images/unittest/unittest-hamburger-menu.png
   :alt: Spyder showing Unittest run button

```

I think the screenshot makes more sense here.

File: doc/plugins/unittest.rst [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+.. image:: /images/unittest/unittest-run-button.png
+   :alt: Spyder showing Unittest run button
+
+You can get an expanded view of the testing messages by clicking on the expand option.
+
+.. image:: /images/unittest/unittest-expanded-message-view.png
+   :alt: Spyder showing Unittest expanded message view
+
+Press on the  :guilabel:`Hamburger Menu Icon` to view different unittest options.
```
Comment (juanis2112) [inline doc/plugins/unittest.rst:65]:
```suggestion

==================
Additional Options
==================

Press on the  :guilabel:`Hamburger Menu Icon` to view different unittest options.
```

Adding a title to the Options menu section.

**Review by juanis2112 — CHANGES_REQUESTED** [review]
Thanks @pkpatricia for working on this. Sorry for getting back so late, I was busy with our new workshops for Spyder 5. Some minor changes that need to be added before merging but this is looking great!

**Comment by pkpatricia** [comment]
added suggested changes

**Comment by CAM-Gerlach** [comment]
Hey @pkpatricia , thanks for all your great work on this, and it looks this is getting close to merge-worthy. Once you've addressed the outstanding feedback and you're ready for another review, please ping me and I can do a pass to check for technical, copyediting and style issues (I'm unclear whether this is ready, as you left a comment saying the suggestions were applied, but then made further commits and I don't see the other reviewers re-requested for review). Most if not all of these should be doable as suggestions and not require any additional work from you other than batch-commiting the changes. If you prefer, given they should all be very minor, I can make them as a commit directly on your branch (minus any that aren't strictly technical in nature).

Speaking of which, when a reviewer makes a number of suggestions in a single review, its usually much better to use Github's "batch" feature to add each commit to commit the changes all at once. From the Files view, you just click "add to batch" on each of the suggestions, and then when they are all selected, and then at the top, add an appropriately descriptive commit message (e.g. `Apply reviewer feedback on Unittest plugin docs` or `Address style issues in Unittest plugin docs`), and then click the `Commit` button. This will avoid lots of trivial commits with duplicate, uninformative commit messages that will need to be fixuped before merging, as well as running our CIs many more times and making it more work to review.

To fix the commits already made, you'll need to run `git rebase -i 5f59c26fde2221d557c2c7e94a13e7f2ee2db0b0`, and change the `pick` next to all but the first (top) commit in each batch to `fixup`, and then save and close the file. Git will do the rebase, and then you'll need to `git push -f` to push them. Alternatively, if you're not comfortable with rewriting git history (a mistake will require more git fu to undo), we can do it ourselves before merging. The fixes/optimizations to the PNG images also needs to be rebased in, which we can also do for you if you prefer.

Thanks!

**Comment by CAM-Gerlach** [comment]
Hey @pkpatricia , not sure if you saw my earlier comment, but is this ready for another pass? Thanks!

Also, just FYI, I marked it as draft as a reminder to us to make sure that the duplicate commits and optimized images are rebased in before irrevocably merging, per our standard procedure.

**Comment by pkpatricia** [comment]
I'm super sorry.  I'm getting the hang of how git pull requests work.  Yes,
I believe this is ready for submission.  When I first started this
process,  Carlos and Juanita instructed me on using `make docs` and
optimizing images. Either way I know that inorder for this to be posted it
must be correct so I'm open to improvement.

Best Regards
"You always have two choices: your commitment versus your fear."
-Sammy Davis, Jr.
https://patriciaparker.co

**Comment by CAM-Gerlach** [comment]
No worries at all @pkpatricia ! You're all good, and we really appreciate your great help and your patience with our feedback. For my part, I just came back from an extended absence for personal reasons, so sorry if I'm not 100% up to date myself. It wasn't too long ago (especially for me and Juanita) that we were new to all this too; I certainly remember it being kinda overwhelming and even intimidating at times, and we're always still learning new things—and at least as for my first pull requests, they certainly weren't as impressive as this. If there's anything confusing or unclear, or something we can do to make this whole process smoother, just ask—we're here to help.

In any case, I'm looking forward to reviewing your work, and will get on that now. Once its ready to merge, its no problem at all for us to take care of cleaning up the commits and rebasing in the scaled/optimized images and GIFs (in fact, its pretty typical). Thanks again!

File: doc/plugins/unittest.rst [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+.. image:: /images/unittest/unittest-hamburger-menu.png
+   :alt: Spyder Unittest hamburger menu in Spyder
+
+The configuration view will show you how to choose the pytest option and the directory of where you want to run the test file(s) from.
+
+.. image:: /images/unittest/unittest-configuration-view.png
+   :alt: Spyder Unittest configuration view in Spyder
+
+The output viewer can be used to directly view the output results of your testing.  This view can be very useful for viewing the results of unittests.
```
Comment (CAM-Gerlach) [inline doc/plugins/unittest.rst:82]:
```suggestion
The :guilabel:`Show output` dialog can be used to directly view the raw output of your tests.
This can be useful to see more details of the problem when a test fails.
```

* Rewrite redundant second sentence to avoid duplicating the content in the first and instead provide unique information
* Use more specific phrasing and teminology
* Avoid word repetition
* Refer to the actual labels shown in the GUI and use appropriate roles
* One sentence per line

File: doc/plugins/unittest.rst [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+.. image:: /images/unittest/unittest-output-viewer.png
+   :alt: Spyder Unittest output viewer in Spyder
+
+Finally, the dependencies view will show you all of your installed testing frameworks.
```
Comment (CAM-Gerlach) [inline doc/plugins/unittest.rst:87]:
```suggestion
Finally, the :guilabel:`Dependencies` dialog will show all of your installed testing frameworks and their versions.
```

* Use more specific phrasing and teminology
* Refer to the actual labels shown in the GUI and use appropriate roles

**Review by CAM-Gerlach — CHANGES_REQUESTED** [review]
Hey @pkpatricia , thanks again for this! Its an excellent start; I just had a number of technical, style and copyediting comments, all but a couple made as GitHub suggestions that you can add to batch and then apply in one go, assuming you're okay with them. 

Your screenshots all pretty much looked great in terms of content; I did notice that they are generally rather sparse, the text/elements are rather small and there isn't a clear focus on what the screenshot intends to show. To fix this in general, I suggest (on all screenshots except where noted):

* [ ] Crop the screenshot appropriately to show just the areas of interest; e.g. crop halway down for the first two, take a screenshot of just the pane for the next three (undocking the pane is useful here), and when showing a dialog, just show the dialog for the last three.
* [ ] Increase the text size in the editor (if shown)
* [ ] Add a few more tests so the Unittest pane isn't so empty
* [ ] Increase your screen's DPI scaling factor to the largest setting (under `System Preferences > Displays` to the setting that makes text and UI elements the largest)

Also, the screenshots will need to be rescaled to half their size via nearest neighbor resampling (to maximize compression efficiency), i.e. double their standard DPI display size, to avoid taking up excessive space, and the change rebased in. Given this would only take me a moment to do and it requires some advanced Git manipulation that can lead to problems if you're not careful, and the PR needs a rebase anyway to fixup the lots of duplicate commits, don't worry about this part and just commit the screenshots full size, and I'll take care of that before merging.

Thanks!

File: doc/plugins/unittest.rst [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+###############
+
+**Spyder-unittest** is a plugin that integrates popular unit test frameworks with Spyder: such as Pytest, Unittest and Nose. It allows you to run test suites and view the results in the IDE.
+
+.. image:: /images/unittest/unittest-run.png
+   :alt: Spyder Unittest in Spyder
+
+| `Pytest <https://docs.pytest.org/en/stable/getting-started.html#install-pytest>`_
+
+| `Unittest <https://docs.python.org/3/library/unittest.html#module-unittest>`_
```
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply doc/plugins/unittest.rst]: I believe this was superceded by another suggestion of yours.

File: doc/plugins/unittest.rst (hunk 11 above)
Comment (CAM-Gerlach) [reply doc/plugins/unittest.rst]: Looks like this was fixed elsewhere

**Comment by CAM-Gerlach** [comment]
Hey @pkpatricia , I'm a little confused as to what happened...is something wrong? All this PR needed to be ready to go was batch-applying the requested copyediting, technical and style suggestions, addressing a couple previous reviewer comments and potentially making some changes to the screenshots, if the other reviewers agree and that's not too much work (otherwise, we can take care of that later). I again apologize if my technical/style/copyediting pass was rather delayed; I was unfortunately taking a break from Spyder due to the passing of several people close to me. We're here to help get this across the finish line for you, and if you feel differently about any of the suggested changes, please let us know and we'll work something out. Hoping to hear from you soon.
